### ISSUE TITLE
notebook 5, pos_embeddings, magic number 100

### ISSUE THREAD

**Issue by antgr** [bentrevett/pytorch-seq2seq#32]
In encoder and decoder, we define pos_embeddings using size 100
`self.pos_embedding = nn.Embedding(100, emb_dim)`
Why we chose 100? Is it the maximum size of a sentence for all batches?

**Comment by bentrevett** [comment]
Just note that notebook 5 is still in (very slow) progress, so it is not completely finished. 

In theory the "vocab size" (the magic 100) for the positional embedding should be equal to the maximum sentence length, yes. This is because we need an entry in the positional embedding for all possible positions. 

However, what happens if we set it to the maximum length sentence in our training/validation/test set (say 100) and then when it comes to inference we try and use a sentence with 150 tokens? We'll get an error because the "vocab size" (the number of rows in the embedding layer weights) will only be 100. 

So, should we make it larger than 100? Well, if we did then we wouldn't ever actually learn the positional embeddings for positions greater than 100 - as we have no sentences with over 100 tokens - and these embeddings would just be random. 

So I believe in practice, when doing inference, the common thing to do is make that 100 equal to the maximum length of the sentence in the **training set** (as if we had an example in the validation set with 120 tokens we wouldn't update the positional embeddings on it anyway) and then during validation/test/inference just cut the sentence down to only use the first 100 tokens.

This is a very long roundabout way of saying: yes, you're correct - it should be the maximum length of the sentences in our training set. Once I get around to finishing notebook 5 I'll be sure to actually code it to use the maximum training sentence length and add some comments.
